sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -684,6 +685,353 @@ public virtual async Task CloseAsync(CancellationToken cancellationToken = defau
+
+                    // Update state and commit the sequencing.
+
+                    EventHubsEventSource.Log.IdempotentSequenceUpdate(EventHubName, options.PartitionId, partitionState.LastPublishedSequenceNumber.Value, lastSequence);
+                    partitionState.LastPublishedSequenceNumber = lastSequence;
+
+                    foreach (var eventData in eventSet)
+                    {
+                        eventData.CommitPublishingSequenceNumber();
+                    }
+                }
+                catch
+                {
+                    // Clear the pending sequence numbers in the face of an exception.
+
+                    foreach (var eventData in eventSet)
+                    {
+                        eventData.PendingPublishSequenceNumber = null;
+                    }
+
+                    throw;
+                }
+                finally
+                {
+                    partitionState.PublishingGuard.Release();
+                    EventHubsEventSource.Log.IdempotentSynchronizationRelease(EventHubName, options.PartitionId);
+                }
+            }
+            catch (Exception ex)
+            {
+                EventHubsEventSource.Log.IdempotentPublishError(EventHubName, options.PartitionId, ex.Message);
+                throw;
+            }
+            finally
+            {
+
+                EventHubsEventSource.Log.IdempotentPublishComplete(EventHubName, options.PartitionId);
+            }
+        }
+
+        /// <summary>
+        ///   Sends a set of events to the associated Event Hub using a batched approach.
+        /// </summary>
+        ///
+        /// <param name="eventBatch">The set of event data to send. A batch may be created using <see cref="CreateBatchAsync(CancellationToken)" />.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken" /> instance to signal the request to cancel the operation.</param>
+        ///
+        private async Task SendIdempotentAsync(EventDataBatch eventBatch,
+                                               CancellationToken cancellationToken = default)
+        {
+            var options = eventBatch.SendOptions;
+
+            AssertPartitionIsReferenced(options.PartitionId);
+            AssertIdempotentBatchNotPublished(eventBatch);
+
+            try
+            {
+                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+                EventHubsEventSource.Log.IdempotentPublishStart(EventHubName, options.PartitionId);
+
+                var partitionState = PartitionState.GetOrAdd(options.PartitionId, new PartitionPublishingState(options.PartitionId));
+
+                var eventSet = eventBatch.AsEnumerable<EventData>() switch
+                {
+                    IReadOnlyList<EventData> eventList => eventList,
+                    IEnumerable<EventData> eventEnumerable => eventEnumerable.ToList()
+                };
```

Comment:
Unfortunately, it always allocates a new list when you `ToList()` it.   In the implementation we're 90% sure that we're getting back a list in most cases.   It may even make sense to make this into an `AsList` at some point, but I'll punt that for now.

Reply:
I benchmarked this just out of curiosity - you are 100% correct 😄 

```
|               Method |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|
--------------------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
|          AlreadyList | 29.739 ns | 0.1754 ns | 0.1641 ns | 0.0092 |     - |     - |      40 B |
|           Enumerable | 67.256 ns | 1.3537 ns | 2.0672 ns | 0.0157 |     - |     - |      68 B |
|    SwitchAlreadyList |  1.824 ns | 0.0098 ns | 0.0076 ns |      - |     - |     - |         - |
| SwitchNotAlreadyList | 39.886 ns | 0.1303 ns | 0.1088 ns | 0.0139 |     - |     - |      60 B |
```